src/Columns/ColumnDynamic.cpp

```diff
@@ -682,64 +1019,84 @@ void ColumnDynamic::takeDynamicStructureFromSourceColumns(const Columns & source
+        result_variants.reserve(global_max_dynamic_types + 1);
+        for (const auto & [size, variant] : variants_with_sizes)
         {
-            if (result_variants.size() == max_dynamic_types)
-                break;
-
-            if (variant->getName() != "String")
+            /// Add variant to the resulting variants list until we reach max_dynamic_types.
+            if (result_variants.size() < global_max_dynamic_types)
                 result_variants.push_back(variant);
+            /// Add all remaining variants into shared_variants_statistics until we reach its max size.
+            else if (new_statistics.shared_variants_statistics.size() < Statistics::MAX_SHARED_VARIANT_STATISTICS_SIZE)
+                new_statistics.shared_variants_statistics[variant->getName()] = size;
+            else
+                break;
         }
 
+        /// Add shared variant.
+        result_variants.push_back(getSharedVariantDataType());
         result_variant_type = std::make_shared<DataTypeVariant>(result_variants);
     }
     else
     {
         result_variant_type = std::make_shared<DataTypeVariant>(all_variants);
     }
 
-    /// Now we have resulting Variant and can fill variant info.
-    variant_info.variant_type = result_variant_type;
-    variant_info.variant_name = result_variant_type->getName();
-    const auto & result_variants = assert_cast<const DataTypeVariant &>(*result_variant_type).getVariants();
-    variant_info.variant_names.clear();
-    variant_info.variant_names.reserve(result_variants.size());
-    variant_info.variant_name_to_discriminator.clear();
-    variant_info.variant_name_to_discriminator.reserve(result_variants.size());
-    statistics.data.clear();
-    statistics.data.reserve(result_variants.size());
-    statistics.source = Statistics::Source::MERGE;
-    for (size_t i = 0; i != result_variants.size(); ++i)
-    {
-        auto variant_name = result_variants[i]->getName();
-        variant_info.variant_names.push_back(variant_name);
-        variant_info.variant_name_to_discriminator[variant_name] = i;
-        statistics.data[variant_name] = total_sizes[variant_name];
-    }
+    /// Now we have resulting Variant and can fill variant info and create merge statistics.
+    setVariantType(result_variant_type);
+    new_statistics.variants_statistics.reserve(variant_info.variant_names.size());
+    for (const auto & variant_name : variant_info.variant_names)
+        new_statistics.variants_statistics[variant_name] = total_sizes[variant_name];
+    statistics = std::make_shared<const Statistics>(std::move(new_statistics));
 
-    variant_column = variant_info.variant_type->createColumn();
+    /// Reduce max_dynamic_types to the number of selected variants (without shared variant), so there will be no possibility
+    /// to extend selected variants on inerts into this column during merges.
+    max_dynamic_types = variant_info.variant_names.size() - 1;
```

Comment:
now we decrease max_dynamic_types by 1 instantly, how will it affect other compares with `- 1`?

Reply:
It will not affect other compares. `max_dynamic_types` always contains the limit on dynamic types withouy shared variant and here we set it to the number of dynamic types without shared variant.

So for example, if we have dynamic types `String`, `UInt32` and `Array(UInt32)` `variant_info.variant_names.size()` will be `4` (3 types + shared variant) and we will set the limit to `3`. And we check if limit is reached by checking `variant_info.variant_names.size() - 1 == max_dynamic_types` and in this example it will be always `4 - 1 == 3`. So it's correct.